### ISSUE TITLE
performance of using multiple nodes

### ISSUE THREAD

**Issue by hfutxrg** [dmlc/mxnet#3422]
I have used multiple GPUs within one node and got good scalability. But When I use multiple nodes, it does not scale well. Does anyone know why?  Let me explain in detail.

I was using inception-bn network to train the imagenet data. The speed is around 
187 samples/sec with 1 GPU, batch-size=64
373 samples/sec with 2 GPUs, batch-size=128
741 samples/sec with 4 GPUs. batch-size=256

Now I am using two nodes where each node uses 4 GPUs. I expect that in each node, the speed is 741 samples/sec. But the reality is 
on node 1, the speed is 741 samples/sec
but on node 2 the speed is only around 338 samples/sec. 

I was using the train_imagenet.py in examples/images-classcification. My launch command is:
../../tools/launch.py -H hosts -n 2 --launcher ssh python train_imagenet.py --network inception-bn --gpus 0,1,2,3 --batch-size 256 --num-epochs 1 --data-shape 224 --data-dir ../../data/ilsvrc12/ --kv-store dist_async.

Both nodes have 1 ethernet address and 1 infiniband address. 
The hosts file contains the infiniband addresses of two nodes:
10.149.0.1
10.149.0.2

I ran the above command from 10.149.0.1. Although I specified the infiniband addresses in hosts file, the log file still shows the program is using the ethernet address. So I want to know is the poor performance on node 2 because of using the ethernet for communication, or any other reasons? Should I change the batch size when using multiple nodes?

**Comment by piiswrong** [comment]
try sync sgd

**Comment by hfutxrg** [comment]
I tried sync, but the performance was worse.
I tried the following combinations:
1. kv-store=dist_sync
No matter what addresses are specified in hosts file (ethernet or infiniband addresses), both nodes node[0] and node[1] have the speed ~343 sample/sec

2. kv-store=dist_async
No matter what addresses are specified in hosts file (ethernet or infiniband addresses), node[0] has the speed 740 samples/sec,  node[1] has the speed ~343 sample/sec

No matter from what node I executed the launch.py, the performance on node[0] and node[1] is the same as the performance I executed from another node.
So does it mean if I execute the program from node001, node[0] is node001 and node[1] is node002.
and if I execute the program from node002, node[0] is node002 and node[1] is node001?

I tried to let the program use infiniband address. As I asked in another question, you told me to put "export DMLC_INTERFACCE=ib0" between launch.py and training script:
../../tools/launch.py -H hosts -n 2 --launcher ssh export DMLC_INTERFACE=ib0; python     train_imagenet.py --network inception-bn --gpus 0,1,2,3 --batch-size 256 --num-epochs 1 --data-shape 224 --data-dir ../../data/ilsvrc12/ --kv-store dist_async

then it has the following error:
mxnet/dmlc-core/include/dmlc/logging.h:235: [12:26:17] src/postoffice.cc:16: Check  notnull: Environment::Get()->find("DMLC_NUM_WORKER")

Then I tried to specify DMLC_NUM_WORKER, but it asked for more environment variables. So I specified all of them:
../../tools/launch.py -H hosts -n 2 --launcher ssh export DMLC_INTERFACE=ib0; export DMLC_NUM_WORKER=2; export DMLC_NUM_SERVER=2; export DMLC_ROLE=worker; python     train_imagenet.py --network inception-bn --gpus 0,1,2,3 --batch-size 256 --num-epochs 1 --data-shape 224 --data-dir ../../data/ilsvrc12/ --kv-store dist_async

but again has the following error:
terminate called after throwing an instance of 'std::logic_error'
  6   what():  basic_string::_S_construct null not valid
  7 ./run.sh: line 18: 36014 Aborted                 python train_imagenet.py --network inception-bn --gpus 0,1,2,3 --batch-size 256 --num-epochs 1 --data-shape 224 --data    -dir ../../data/ilsvrc12/ --kv-store dist_async

The errors are the same if I use dist_sync instead of dist_async.

So how to utilize another network interface correctly and how to get the scalable performance when using multiple nodes?

**Comment by hfutxrg** [comment]
I have found the issue was because it could not use infiniband interface. After I fixed this issue, then the performance scales very well. MXNet is an excellent framework.

**Comment by piiswrong** [comment]
Would you mind sharing how you fixed it? It would be greatly appreciated if you can send a pull request.

**Comment by hfutxrg** [comment]
I explained part of my solution in the post https://github.com/dmlc/mxnet/issues/3400. I will put them here again. 
It seems the issue was that MXNet could not get the values of those environment variables correctly. So I hard-coded those environment variable values in the sourcecode. 

I made two changes, both are in function "void Van::Start()" in mxnet/ps-lite/src/van.cc:
the line
scheduler_.hostname = std::string(CHECK_NOTNULL(Environment::Get()->find("DMLC_PS_ROOT_URI")));
was changed to 
scheduler_.hostname = std::string("10.149.0.1"); // this is the infiniband ip of my scheduler node

and the line 
if (itf) interface = std::string(itf);
was changed to 
interface = std::string("ib0"); // the infiniband interface on all of my worker nodes

because it seems the line "const char*  itf = Environment::Get()->find("DMLC_INTERFACE");" was also not working. The value of variable interface was empty when I printed it out.

So, although I set the environment variable DMLC_INTERFACE and DMLC_PS_ROOT_URI, it seems MXNet could not get their values correctly, at least it was not working on my machine. I solved it in a non-elegant way. If it is a bug, maybe you can fix it and push the changes into the repo.

**Comment by yhpeng-git** [comment]
I also have this bug in Ubuntu 14.04.5 now. It seems that it has not been fixed yet.

**Comment by formath** [comment]
As I know, the environment variables can't be transmitted to remote nodes.
